prow/spyglass/gcsartifact.go

```diff
@@ -0,0 +1,204 @@
+		return 0, viewers.ErrGzipOffsetRead
+	}
+	reader, err := a.handle.NewRangeReader(a.ctx, off, int64(len(p)))
+	defer reader.Close()
+	if err != nil {
+		return 0, fmt.Errorf("error getting artifact reader: %v", err)
+	}
+	n, err = reader.Read(p)
+	if err != nil {
+		return 0, fmt.Errorf("error reading from artifact: %v", err)
+	}
+	return n, nil
+}
+
+// ReadAtMost reads at most n bytes from a file in GCS. If the file is compressed (gzip) in GCS, n bytes
+// of gzipped content will be downloaded and decompressed into potentially GREATER than n bytes of content.
+func (a *GCSArtifact) ReadAtMost(n int64) ([]byte, error) {
+	reader, err := a.handle.NewRangeReader(a.ctx, 0, n)
+	defer reader.Close()
+	if err != nil {
+		return nil, fmt.Errorf("error getting artifact reader: %v", err)
+	}
+	var p []byte
+	var e error
+	gzipped, err := a.gzipped()
+	if err != nil {
+		return nil, fmt.Errorf("error checking artifact for gzip compression: %v", err)
+	}
+	if gzipped {
+		gReader, err := gzip.NewReader(reader)
+		if err != nil {
+			return nil, fmt.Errorf("error getting gzip reader: %v", err)
+		}
+		p, e = ioutil.ReadAll(gReader)
+		if e != nil {
+			return nil, fmt.Errorf("error reading all from gzipped artifact: %v", err)
+		}
+	} else {
+		p, e = ioutil.ReadAll(reader)
+		if e != nil {
+			return nil, fmt.Errorf("error reading all from artifact: %v", err)
+		}
+	}
+	return p, nil
+}
+
+// ReadAll will either read the entire file or throw an error if file size is too big
+func (a *GCSArtifact) ReadAll() ([]byte, error) {
+	size, err := a.Size()
+	if err != nil {
+		return nil, fmt.Errorf("error getting artifact size: %v", err)
+	}
+	if size > a.sizeLimit {
+		return nil, viewers.ErrFileTooLarge
+	}
+	reader, err := a.handle.NewReader(a.ctx)
+	defer reader.Close()
+	if err != nil {
+		return nil, fmt.Errorf("error getting artifact reader: %v", err)
+	}
+	p, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return nil, fmt.Errorf("error reading all from artifact: %v", err)
+	}
+	return p, nil
+}
+
+// ReadTail reads the last n bytes from a file in GCS
+func (a *GCSArtifact) ReadTail(n int64) ([]byte, error) {
+	gzipped, err := a.gzipped()
+	if err != nil {
+		return nil, fmt.Errorf("error checking artifact for gzip compression: %v", err)
+	}
+	if gzipped {
+		return nil, viewers.ErrGzipOffsetRead
+	}
+	size, err := a.Size()
+	if err != nil {
+		return nil, fmt.Errorf("error getting artifact size: %v", err)
+	}
+	var offset int64
+	if n > size {
+		offset = 0
+	} else {
+		offset = size - n
+	}
+	reader, err := a.handle.NewRangeReader(a.ctx, offset, -1)
+	defer reader.Close()
+	if err != nil {
+		return nil, fmt.Errorf("error getting artifact reader: %v", err)
+	}
+	read, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return nil, fmt.Errorf("error reading all from artiact: %v", err)
+	}
+	return read, nil
+}
+
+// UseContext sets the context to be used in GCS read operations
+func (a *GCSArtifact) UseContext(ctx context.Context) error {
```

Comment:
Are artifacts local to a request? This seems odd, should be set at creation time

Reply:
It is already set at creation time. This is for use by viewers who may want a cancellation context

Reply:
We could just always provide a cancelation context in case a viewer wants to cancel.

Reply:
The point there being that there are context-specific features the viewer might want, and we dont want to reimplement context functionality on every Artifact.